File: TheWatch-iOS/ViewModels/TwoFactorViewModel.swift
// TwoFactorViewModel.swift
// TheWatch-iOS
//
// ViewModel for the two-factor authentication verification screen.
// Shown after login when the user's account has MFA enabled.
//
// Supports three MFA methods:
//   - TOTP (authenticator app such as Google Authenticator, Authy, 1Password)
//   - SMS  (6-digit code sent to the user's registered phone number)
//   - Backup code (one-time 8-character recovery code)
//
// Example flow:
//   1. User logs in with email/password.
//   2. Login succeeds but backend returns isMfaEnabled = true.
//   3. App navigates to TwoFactorView with this ViewModel.
//   4. User enters 6-digit code from their authenticator or SMS.
//   5. ViewModel calls authService.verifyMfaCode().
//   6. On success, navigation proceeds to HomeView.
//
// The SMS resend cooldown prevents users from spamming the resend button.
// Default cooldown is 60 seconds, matching industry standard (NIST SP 800-63B).
//
// Write-Ahead Log: Every verification attempt (success or failure) should be
// logged by the auth service for audit trail compliance.

import Foundation
import Combine

@MainActor
@Observable
final class TwoFactorViewModel {

    // ── Input ────────────────────────────────────────────────────────
    /// Individual digits for the 6-digit code entry.
    /// Each element corresponds to one TextField in the UI.
    var digits: [String] = Array(repeating: "", count: 6)

    /// The currently selected MFA method.
    var selectedMethod: MfaMethod = .totp

    // ── State ────────────────────────────────────────────────────────
    var isLoading = false
    var errorMessage: String?
    var isVerified = false

    /// Seconds remaining before the user can resend an SMS code.
    var resendCooldownSeconds: Int = 0

    /// True when the SMS resend cooldown is active.
    var isResendDisabled: Bool {
        resendCooldownSeconds > 0
    }

    // ── Dependencies ─────────────────────────────────────────────────
    private let authService: any AuthServiceProtocol
    private var cooldownTimer: Timer?

    // MARK: - MFA Method Enum

    enum MfaMethod: String, CaseIterable, Identifiable {
        case totp = "Authenticator App"
        case sms = "SMS"
        case backup = "Backup Code"

        var id: String { rawValue }

        /// The raw string expected by the backend API.
        var apiValue: String {
            switch self {
            case .totp: return "totp"
            case .sms: return "sms"
            case .backup: return "backup"
            }
        }

        /// SF Symbol name for the method icon.
        var iconName: String {
            switch self {
            case .totp: return "lock.shield"
            case .sms: return "message"
            case .backup: return "key"
            }
        }

        /// Maximum code length for this method.
        var codeLength: Int {
            switch self {
            case .totp, .sms: return 6
            case .backup: return 8
            }
        }
    }

    // MARK: - Computed

    /// The full code assembled from individual digit fields.
    var fullCode: String {
        digits.joined()
    }

    /// True when enough digits have been entered for the selected method.
    var isCodeComplete: Bool {
        let required = selectedMethod.codeLength
        let code = fullCode.trimmingCharacters(in: .whitespaces)
        return code.count >= required
    }

    // MARK: - Init

    init(authService: any AuthServiceProtocol) {
        self.authService = authService
    }

    deinit {
        cooldownTimer?.invalidate()
    }

    // MARK: - Actions

    /// Verify the entered code against the backend.
    func verify() async {
        guard isCodeComplete else { return }

        isLoading = true
        errorMessage = nil

        do {
            let result = try await authService.verifyMfaCode(
                code: fullCode,
                method: selectedMethod.apiValue
            )

            if result {
                isVerified = true
            } else {
                errorMessage = "Invalid code. Please try again."
                clearCode()
            }
        } catch {
            errorMessage = error.localizedDescription
            clearCode()
        }

        isLoading = false
    }

    /// Resend an SMS verification code.
    /// Only applicable when selectedMethod == .sms.
    func resendSmsCode() async {
        guard selectedMethod == .sms, !isResendDisabled else { return }

        isLoading = true
        errorMessage = nil

        do {
            // The backend resend endpoint is the same as MFA verify with a "resend" flag.
            // We call verifyMfaCode with a special empty code to trigger resend.
            // Alternatively, the enrollMfa method can be used to re-trigger SMS.
            _ = try await authService.enrollMfa(method: "sms", phoneNumber: nil)
            startResendCooldown()
        } catch {
            errorMessage = "Failed to resend code. Please try again."
        }

        isLoading = false
    }

    /// Switch the MFA method and reset the code entry.
    func switchMethod(to method: MfaMethod) {
        selectedMethod = method
        clearCode()
        errorMessage = nil

        // Adjust digit count for backup codes
        if method == .backup {
            digits = Array(repeating: "", count: 8)
        } else {
            digits = Array(repeating: "", count: 6)
        }
    }

    /// Clear all entered digits.
    func clearCode() {
        let count = selectedMethod == .backup ? 8 : 6
        digits = Array(repeating: "", count: count)
    }

    /// Handle digit entry with auto-advance.
    /// Called when a digit field changes. Trims to 1 character and advances focus.
    /// - Returns: The index of the next field to focus, or nil if this was the last.
    func handleDigitEntry(at index: Int, newValue: String) -> Int? {
        // Only keep the last character entered (handles paste of full code)
        if newValue.count > 1 {
            // User pasted a full code
            let code = newValue.prefix(digits.count)
            for (i, char) in code.enumerated() {
                digits[i] = String(char)
            }
            return min(code.count, digits.count - 1)
        }

        // Filter non-alphanumeric for backup codes, non-digit for TOTP/SMS
        if selectedMethod == .backup {
            digits[index] = String(newValue.prefix(1))
        } else {
            let filtered = newValue.filter { $0.isNumber }
            digits[index] = String(filtered.prefix(1))
        }

        // Auto-advance to next field
        if !digits[index].isEmpty && index < digits.count - 1 {
            return index + 1
        }

        return nil
    }

    /// Handle backspace: if current field is empty, move focus back.
    /// - Returns: The index of the previous field to focus, or nil.
    func handleBackspace(at index: Int) -> Int? {
        if digits[index].isEmpty && index > 0 {
            digits[index - 1] = ""
            return index - 1
        }
        return nil
    }

    // MARK: - Private

    private func startResendCooldown() {
        resendCooldownSeconds = 60

        cooldownTimer?.invalidate()
        cooldownTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] timer in
            Task { @MainActor in
                guard let self else {
                    timer.invalidate()
                    return
                }
                self.resendCooldownSeconds -= 1
                if self.resendCooldownSeconds <= 0 {
                    timer.invalidate()
                    self.resendCooldownSeconds = 0
                }
            }
        }
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/ui/screens/twofactor/TwoFactorViewModel.kt
package com.thewatch.app.ui.screens.twofactor

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.thewatch.app.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * ViewModel for the Two-Factor Authentication verification screen.
 *
 * Shown after login when the user's account has MFA enabled.
 * Supports three verification methods:
 *   - TOTP: 6-digit code from an authenticator app (Google Authenticator, Authy, etc.)
 *   - SMS:  6-digit code sent via text message to the user's registered phone
 *   - Backup: One-time recovery code (8-10 alphanumeric characters)
 *
 * Write-Ahead Log:
 *   - WAL Entry: MFA_SCREEN_OPEN        -> userId, timestamp
 *   - WAL Entry: MFA_METHOD_SWITCH       -> fromMethod, toMethod, timestamp
 *   - WAL Entry: MFA_CODE_SUBMIT         -> method, codeLength, timestamp
 *   - WAL Entry: MFA_VERIFY_SUCCESS      -> method, timestamp
 *   - WAL Entry: MFA_VERIFY_FAIL         -> method, errorMsg, timestamp
 *   - WAL Entry: MFA_SMS_RESEND          -> timestamp
 *   - WAL Entry: MFA_SMS_COOLDOWN_EXPIRE -> timestamp
 *
 * Example usage:
 *   // In navigation, after login returns user.mfaEnabled == true:
 *   navController.navigate(NavRoute.TwoFactor.route)
 *
 *   // In TwoFactorScreen composable:
 *   val viewModel: TwoFactorViewModel = hiltViewModel()
 *   viewModel.verifyCode("123456")
 */
@HiltViewModel
class TwoFactorViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow<TwoFactorUiState>(TwoFactorUiState.Idle)
    val uiState: StateFlow<TwoFactorUiState> = _uiState.asStateFlow()

    private val _selectedMethod = MutableStateFlow(MfaMethod.TOTP)
    val selectedMethod: StateFlow<MfaMethod> = _selectedMethod.asStateFlow()

    private val _smsResendCooldown = MutableStateFlow(0)
    val smsResendCooldown: StateFlow<Int> = _smsResendCooldown.asStateFlow()

    private var countdownJob: Job? = null

    /**
     * Verify the user-entered MFA code against the backend.
     * On success, the backend marks the session as fully authenticated.
     */
    fun verifyCode(code: String) {
        if (code.isEmpty()) {
            _uiState.value = TwoFactorUiState.Error("Please enter the verification code.")
            return
        }

        val method = _selectedMethod.value
        if (method != MfaMethod.BACKUP && code.length != 6) {
            _uiState.value = TwoFactorUiState.Error("Code must be 6 digits.")
            return
        }

        viewModelScope.launch {
            _uiState.value = TwoFactorUiState.Loading
            val result = authRepository.verifyMfaCode(
                code = code,
                method = method.apiValue
            )
            result.fold(
                onSuccess = {
                    _uiState.value = TwoFactorUiState.Success
                },
                onFailure = { e ->
                    _uiState.value = TwoFactorUiState.Error(
                        e.message ?: "Verification failed. Please try again."
                    )
                }
            )
        }
    }

    /**
     * Switch between MFA methods (TOTP, SMS, Backup).
     * If switching to SMS, triggers an SMS send with cooldown.
     */
    fun selectMethod(method: MfaMethod) {
        _selectedMethod.value = method
        _uiState.value = TwoFactorUiState.Idle
        if (method == MfaMethod.SMS && _smsResendCooldown.value == 0) {
            resendSmsCode()
        }
    }

    /**
     * Request the backend to resend an SMS verification code.
     * Starts a 60-second cooldown to prevent abuse.
     */
    fun resendSmsCode() {
        if (_smsResendCooldown.value > 0) return

        viewModelScope.launch {
            _uiState.value = TwoFactorUiState.Loading
            val result = authRepository.verifyMfaCode(
                code = "RESEND",
                method = "sms_resend"
            )
            result.fold(
                onSuccess = {
                    _uiState.value = TwoFactorUiState.SmsSent
                    startResendCooldown()
                },
                onFailure = { e ->
                    _uiState.value = TwoFactorUiState.Error(
                        e.message ?: "Failed to resend SMS. Try again."
                    )
                }
            )
        }
    }

    private fun startResendCooldown() {
        countdownJob?.cancel()
        _smsResendCooldown.value = 60
        countdownJob = viewModelScope.launch {
            while (_smsResendCooldown.value > 0) {
                delay(1000)
                _smsResendCooldown.value -= 1
            }
        }
    }

    fun resetUiState() {
        _uiState.value = TwoFactorUiState.Idle
    }
}

/**
 * MFA verification methods.
 * Each maps to an API method string used by POST /api/account/mfa/verify.
 */
enum class MfaMethod(val apiValue: String, val displayName: String) {
    TOTP("totp", "Authenticator App"),
    SMS("sms", "Text Message"),
    BACKUP("backup", "Backup Code")
}

sealed class TwoFactorUiState {
    object Idle : TwoFactorUiState()
    object Loading : TwoFactorUiState()
    object Success : TwoFactorUiState()
    object SmsSent : TwoFactorUiState()
    data class Error(val message: String) : TwoFactorUiState()
}
